integration/integration_test_base.py

```diff
@@ -0,0 +1,151 @@
+"""@package integration_test_base
+Base classes for Nighthawk integration tests
+"""
+
+import json
+import logging
+from nighthawk_test_server import NighthawkTestServer
+import os
+import threading
+import time
+import subprocess
+import sys
+import socket
+import unittest
+
+
+class IntegrationTestBase(unittest.TestCase):
+  """
+    IntegrationTestBase facilitates testing against the Nighthawk test server, by running
+    determining a free port, and starting it up in a separate process in setUp().
+    """
+
+  def __init__(self, *args, **kwargs):
+    super(IntegrationTestBase, self).__init__(*args, **kwargs)
+    self.test_rundir = os.path.join(os.environ["TEST_SRCDIR"], os.environ["TEST_WORKSPACE"])
+
+    self.nighthawk_test_server_path = os.path.join(self.test_rundir, "nighthawk_test_server")
+    self.nighthawk_test_config_path = None
+    self.nighthawk_client_path = os.path.join(self.test_rundir, "nighthawk_client")
+    self.server_ip = "::1" if IntegrationTestBase.ipv6 else "127.0.0.1"
+    self.socket_type = socket.AF_INET6 if IntegrationTestBase.ipv6 else socket.AF_INET
+    self.test_server = None
+    self.server_port = -1
+    self.parameters = dict()
+
+  def getFreeListenerPortForAddress(self, address):
+    """
+        Determines a free port and returns that. Theoretically it is possible that another process
+        will steal the port before our caller is able to leverage it, but we take that chance.
+        The upside is that we can push the port upon the server we are about to start through configuration
+        which is compatible accross servers.
+        """
+    sock = socket.socket(self.socket_type, socket.SOCK_STREAM)
```

Comment:
Is there a RAII-style `with` patterns that can autoclose the `socket` here?

Reply:
When we move to python 3 we can do that. (see https://github.com/envoyproxy/nighthawk/pull/89/commits/0feec9a30c7fa3b7710f6876cd65e7ef29e145da, but had to back that out)

Reply:
OK, please leave a TODO, I'm keen to get to Python 3 :)

Reply:
done